app/components/osc-webhook-triggers/osc-webhook-triggers.component.js

```diff
@@ -95,42 +127,19 @@
       checkDuplicates(trigger);
     };
 
-    var addEmptyWebhookTrigger = function() {
-      ctrl.webhookTriggers.push({
-        lastTriggerType: "",
-        data: {
-          type: ""
-        }
-      });
-    };
-
-    // Check last trigger if it's type and secret are selected.
-    ctrl.checkLastAndAddNew = function() {
-      var lastTrigger = _.last(ctrl.webhookTriggers);
-      var lastTriggerSecretData = $filter('getWebhookSecretData')(lastTrigger);
-      if (lastTrigger.data.type && (_.has(lastTriggerSecretData, 'secret') || _.has(lastTriggerSecretData, 'secretReference.name'))) {
-        addEmptyWebhookTrigger();
+    ctrl.missingInputError = function(selectBoxName, rowIndex, type, secretReferenceName) {
```

Comment:
It's better if we can use the `ng-*` attributes for this. For instance, set `ng-required` on the secret ref if the trigger type is set or `ng-required` on the trigger type if secret ref is set. I don't think these checks need JS.

If you *do* need to set form validity in JS, we should call `$ctrl.formName.fieldName.$setValidity` and check when the value changes. For example,

https://github.com/openshift/origin-web-console/blob/832cec030ff015081965dd6675e5b0fe98ec02c2/app/scripts/directives/editRequestLimit.js#L117-L118

But it's better to avoid if not necessary.

Reply:
Originally the logic was in the `ng-*` attributes, but based on [conversation](https://github.com/openshift/origin-web-console/pull/2689#discussion_r162735611) we switched to JS, avoid dups in the html template, since for the `ng-class` needs to add/rm the `has-error` class and similar case was for the `ng-show` for the help blocks under the selectboxes.

Since we want to have similar experience then with the EnvVar editor, in which we dont block the `Save` btn if the any of the selectboxes is not selected. Only WH triggers that have selected type and secret will be saved.

Reply:
The original code is also different from what I had in mind. I think it might be straightforward to use the ng-required attributes on the select inputs themselves. Let me follow up.

> Since we want to have similar experience then with the EnvVar editor, in which we dont block the Save btn if the any of the selectboxes is not selected. Only WH triggers that have selected type and secret will be saved.

I don't agree with this behavior either :) If one of the inputs is filled in and the other is not, we should block save. That's invalid, and users might think they're saving changes they're not. I'd call this is a bug in the env editor if it's the current behavior. We should only allow save if both selects are empty (or both have values).

Reply:
@spadgett I understand your concerns and they make sense. I've discussed the design with @jwforres and @ncameronbritt and this what we align on. 
So I guess that before I rewrite the behavior we discuss whats the the proper way the editors should behave.

Reply:
I think there is some miscommunication happening in this thread :)
My opinion:
1) If there is an empty pair (or more than one pair) of inputs, it should not block Save
2) If a pair has partially filled out data, it should block Save. If either field is filled out but the other one is not, then it should not be possible to Save.
3) An empty pair should never be showing a warning
4) A partially filled out pair should not show a warning *until* the entire pair of inputs has lost focus.  This is to avoid filling out the first input, tabbing to the second input, and seeing a warning.

Notice I called out 2 & 4 separately. That is because as soon as you filled out the first input, you should block Save, even though you haven't warned the user yet.

Reply:
+1, this is also what I was thinking.

And I think we can implement it pretty simply with ng-required in the view without needing any JS in the controller.

Reply:
That was my understanding. Thanks for clarifying, @jwforres.